### PR TITLE
Fix directory permissions when unpacking.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Fixes https://github.com/Homebrew/brew/issues/4530.

### CODE REVIEW

**Comment by ilovezfs** [comment]
I think it would be preferable not to flip it back and forth between modes even if it ends up correct at the end.

File: Library/Homebrew/unpack_strategy.rb [hunk 1]
```diff
@@ -96,7 +96,13 @@ def self.can_extract?(path:, magic_number:)
   private
 
   def extract_to_dir(unpack_dir, basename:)
-    FileUtils.cp_r File.join(path, "."), unpack_dir, preserve: true
+    directory_permissions = File.stat(unpack_dir).mode
+
+    begin
+      FileUtils.cp_r File.join(path, "."), unpack_dir, preserve: true
+    ensure
+      FileUtils.chmod directory_permissions, unpack_dir
```
Comment (ilovezfs) [inline Library/Homebrew/unpack_strategy.rb]: Can this fail on EPERM or similar?

Should this be running the chmod in the no-op case?

**Comment by MikeMcQuaid** [comment]
> I think it would be preferable not to flip it back and forth between modes even if it ends up correct at the end.

Agreed 👍. I think bottle permissions can be respected on extraction but only for the contents of the keg. The keg directories could be created manually and the recursive copy just be for the files within there.

**Comment by reitermarkus** [comment]
> the recursive copy just be for the files within there.

I tried to do it this way, but I don't know how to do this.

`FileUtils.cp_r` will follow symlinks, and `FileUtils.copy_entry`, which doesn't, does not work if the destination already exists.

**Comment by reitermarkus** [comment]
Nevermind, apparently `FileUtils.copy_entry` *does*  work.

**Pull request merged**